### PR TITLE
AVS CVV changes

### PR DESCRIPTION
**Pull request by senthilkumar-muthusamy**
Changes from the PR https://github.com/slogsdon/globalpayments-gateway-provider-for-woocommerce/pull/9 for AVS and CVV decline conditions.

### CODE REVIEW

**Review by slogsdon — CHANGES_REQUESTED** [review]
We need to restore the GP API changes (e.g. 3DSecure support). We should only be adding in this feature, not removing / updating anything else unless it is directly related.

**Comment by MSmedal** [comment]
I tested out AVS declines using Heartland/Portico and test keys and it didn't seem to work. Haven't tried CVV yet

**Review by MSmedal — CHANGES_REQUESTED** [review]

File: src/Gateways/AbstractGateway.php [hunk 1]
```diff
@@ -381,6 +402,27 @@ public function init_form_fields() {
...
+			    ),
+			    'avs_reject_conditions'    => array(
+			        'title'       => __( 'AVS Reject Conditions', 'globalpayments-gateway-provider-for-woocommerce' ),
+			        'type'        => 'multiselect',
+			        'description' => __( 'Choose for which AVS result codes, the transaction must be auto reveresed.'),
+			        'options'     => $this->avs_rejection_conditions(),
+			    ),
+			    'cvn_reject_conditions'    => array(
+			        'title'       => __( 'CVN Reject Conditions', 'globalpayments-gateway-provider-for-woocommerce' ),
+			        'type'        => 'multiselect',
```
[the remark this answers is not in the document]
Comment (senthilkumar-muthusamy) [reply src/Gateways/AbstractGateway.php:424]: updated the select box

File: src/Gateways/AbstractGateway.php [hunk 2]
```diff
@@ -381,6 +402,27 @@ public function init_form_fields() {
...
+			        'type'				=> 'checkbox',
+			        'description'		=> sprintf(
+			            __( 'This will check AVS/CVN result codes and reverse transaction.' )
+			            ),
+			        'default'			=> 'no'
+			    ),
+			    'avs_reject_conditions'    => array(
+			        'title'       => __( 'AVS Reject Conditions', 'globalpayments-gateway-provider-for-woocommerce' ),
+			        'type'        => 'multiselect',
+			        'description' => __( 'Choose for which AVS result codes, the transaction must be auto reveresed.'),
```
[the remark this answers is not in the document]
Comment (senthilkumar-muthusamy) [reply src/Gateways/AbstractGateway.php]: updated the select box

File: src/Gateways/AbstractGateway.php [hunk 3]
```diff
@@ -918,6 +975,39 @@ public function woocommerce_available_payment_gateways( $available_gateways ) {
 		return $available_gateways;
 	}
 
+	public function avs_rejection_conditions()
+	{
+	    return array(
+	        'A'  => 'Address matches, zip No Match',
+	        'N'  => 'Neither address or zip code match',
+	        'R'  => 'Retry - system unable to respond',
```
[the remark this answers is not in the document]
Comment (senthilkumar-muthusamy) [reply src/Gateways/AbstractGateway.php]: updated the list for both AVS and CVV

File: src/Gateways/HeartlandGateway.php [hunk 4]
```diff
@@ -70,7 +70,7 @@ public function get_gateway_form_fields() {
 					__( 'This will display a gift card entry field directly below the credit card entry area.' )
 				),
 				'default'			=> 'no'
-			),
+			),
```
[the remark this answers is not in the document]
Comment (senthilkumar-muthusamy) [reply src/Gateways/HeartlandGateway.php]: removed the unwanted space

File: src/Gateways/HeartlandGateway.php [hunk 5]
```diff
@@ -209,5 +209,5 @@ public function process_payment( $order_id ) {
 			'result'   => $is_successful ? 'success' : 'failure',
 			'redirect' => $is_successful ? $this->get_return_url( $order ) : false,
 		);
-	}
+	}
```
[the remark this answers is not in the document]
Comment (senthilkumar-muthusamy) [reply src/Gateways/HeartlandGateway.php]: removed the unwanted space

**Comment by MSmedal** [comment]
@senthilkumar-muthusamy , can you pull v1.0.1 into your branch? I'll run through a bit more testing after that.

File: src/Gateways/AbstractGateway.php [hunk 6]
```diff
@@ -381,6 +402,31 @@ public function init_form_fields() {
...
 					),
 				),
+			    'check_avs_cvv' => array(
+			        'title'				=> __( 'Check AVS CVN', 'globalpayments-gateway-provider-for-woocommerce' ),
+			        'label'				=> __( 'Check AVS/CVN result codes and reverse transaction.', 'globalpayments-gateway-provider-for-woocommerce' ),
+			        'type'				=> 'checkbox',
+			        'description'		=> sprintf(
+			            __( 'This will check AVS/CVN result codes and reverse transaction.' )
+			            ),
+			        'default'			=> 'no'
```
Comment (MSmedal) [inline src/Gateways/AbstractGateway.php:412]: We'll need to have this functionality enabled by default
```suggestion
			        'default'			=> 'yes'
```

File: src/Gateways/AbstractGateway.php [hunk 7]
```diff
@@ -381,6 +402,31 @@ public function init_form_fields() {
...
+			    ),
+			    'avs_reject_conditions'    => array(
+			        'title'       => __( 'AVS Reject Conditions', 'globalpayments-gateway-provider-for-woocommerce' ),
+			        'type'        => 'multiselect',
+			        'class'       => 'wc-enhanced-select',
+			        'css'         => 'width: 450px',
+			        'description' => __( 'Choose for which AVS result codes, the transaction must be auto reveresed.'),
+			        'options'     => $this->avs_rejection_conditions(),
```
Comment (MSmedal) [inline src/Gateways/AbstractGateway.php:420]: add default values here
```suggestion
			        'options'     => $this->avs_rejection_conditions(),
			        'default'     => array("N", "S", "U", "P", "R", "G", "C", "I"),
```

File: src/Gateways/AbstractGateway.php [hunk 8]
```diff
@@ -381,6 +402,31 @@ public function init_form_fields() {
...
+			    ),
+			    'cvn_reject_conditions'    => array(
+			        'title'       => __( 'CVN Reject Conditions', 'globalpayments-gateway-provider-for-woocommerce' ),
+			        'type'        => 'multiselect',
+			        'class'       => 'wc-enhanced-select',
+			        'css'         => 'width: 450px',
+			        'description' => __( 'Choose for which CVN result codes, the transaction must be auto reveresed.'),
+			        'options'     => $this->cvn_rejection_conditions(),
```
Comment (MSmedal) [inline src/Gateways/AbstractGateway.php:428]: and defaults here too
```suggestion
			        'options'     => $this->cvn_rejection_conditions(),
			        'default'     => array("P", "?", "N"),
```

File: src/Gateways/AbstractGateway.php [hunk 9]
```diff
@@ -852,6 +898,21 @@ protected function handle_response( Requests\RequestInterface $request, Transact
 
 			return false;
 		}
+		
+		//reverse incase of AVS/CVN failure
+		if(!empty($response->transactionReference->transactionId) && !empty($this->check_avs_cvv)){
```
Comment (MSmedal) [inline src/Gateways/AbstractGateway.php:903]: I had to change this line as follows for the logic to work properly during testing
```suggestion
		if(!empty($response->transactionReference->transactionId) && $this->get_option('check_avs_cvv') === 'yes'){
```

**Review by MSmedal — CHANGES_REQUESTED** [review]
After coordinating with Desmond's team about what the defaults should be and a round of testing with the Heartland gateway I came up with a couple more items. I added comments and suggestions

File: src/Gateways/AbstractGateway.php (hunk 6 above)
Comment (senthilkumar-muthusamy) [reply src/Gateways/AbstractGateway.php]: updated the code..

File: src/Gateways/AbstractGateway.php (hunk 7 above)
Comment (senthilkumar-muthusamy) [reply src/Gateways/AbstractGateway.php:420]: default conditions added

File: src/Gateways/AbstractGateway.php (hunk 8 above)
Comment (senthilkumar-muthusamy) [reply src/Gateways/AbstractGateway.php:429]: default consitions added

File: src/Gateways/AbstractGateway.php (hunk 9 above)
Comment (senthilkumar-muthusamy) [reply src/Gateways/AbstractGateway.php]: conditions updated

**Review by MSmedal — APPROVED** [review]

**Review by MSmedal — APPROVED** [review]

**Pull request merged**